library/hkdf.c

```diff
@@ -56,11 +61,13 @@ int mbedtls_hkdf_extract( const mbedtls_md_info_t *md,
                           const unsigned char *ikm, size_t ikm_len,
                           unsigned char *prk )
 {
-    unsigned char null_salt[MBEDTLS_MD_MAX_SIZE] = { '\0' };
-
+    const unsigned char null_salt[MBEDTLS_MD_MAX_SIZE] = { '\0' };
+    size_t hash_len;
```

Comment:
Which compiler complained about `size_t hash_len` inside the start of the `if` block? Declaring new variables at the start of a block is allowed by C89.

Reply:
The original code in PR #1497 was C89 compliant, but I added code inside of the if block before the the declaration of  `size_t hash_len`. 

The codeblock below will give a compiler warning on gcc version 8.1.1

```cc
int mbedtls_hkdf_extract( const mbedtls_md_info_t *md,
                          const unsigned char *salt, size_t salt_len,
                          const unsigned char *ikm, size_t ikm_len,
                          unsigned char *prk )
{
    unsigned char null_salt[MBEDTLS_MD_MAX_SIZE] = { '\0' };
    if( salt == NULL )
    {  
        // New code
        if( salt_len != 0 ) {
             return MBEDTLS_ERR_HKDF_BAD_INPUT_DATA;
        }
        // Declaration after new code
        size_t hash_len;
        hash_len = mbedtls_md_get_size( md );
        if( hash_len == 0 )
        {
            return MBEDTLS_ERR_HKDF_BAD_INPUT_DATA;
        }
        salt = null_salt;
        salt_len = hash_len;
    }
    return( mbedtls_md_hmac( md, salt, salt_len, ikm, ikm_len, prk ) );
}
```
```$ make clean && make                                               
  CC    aes.c
<snip>
  CC    hkdf.c
hkdf.c: In function 'mbedtls_hkdf_extract':
hkdf.c:66:9: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
         size_t hash_len;
         ^~~~~~
```